File: backend/src/api/queue.rs
use axum::{Json, extract::{State, Path}};
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use uuid::Uuid;
use crate::{app::AppState, db::queries, error::{AppResult, AppError}, notifications, notifications::telegram::TelegramEvent, pipeline::trigger, pipeline::trigger::get_setting_or_env};

#[derive(Debug, Deserialize)]
pub struct CreateJobRequest {
    pub imdb_id: String,
    pub media_type: String,
    pub season: Option<i64>,
    pub episode: Option<i64>,
    pub title: Option<String>,
    pub poster_url: Option<String>,
    pub magnet_uri: String,
    pub infohash: String,
    pub torrent_name: String,
    pub file_idx: i64,
    pub file_size_bytes: i64,
}

#[derive(Debug, Serialize)]
pub struct CreateJobResponse {
    pub job_id: String,
    pub status: String,
}

#[derive(Debug, Serialize)]
pub struct QueueListResponse {
    pub processing: Vec<queries::Job>,
    pub queued: Vec<queries::Job>,
    pub completed: Vec<queries::Job>,
    pub failed: Vec<queries::Job>,
}

#[derive(Debug, Serialize)]
pub struct JobDetailResponse {
    pub job: queries::Job,
    pub events: Vec<queries::JobEvent>,
    pub gh_repo: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct RetryResponse {
    pub job_id: String,
    pub status: String,
}

pub async fn create_job(
    State(state): State<Arc<AppState>>,
    Json(body): Json<CreateJobRequest>,
) -> AppResult<Json<CreateJobResponse>> {
    let job_id = Uuid::new_v4().to_string();

    let new_job = queries::NewJob {
        id: job_id.clone(),
        imdb_id: body.imdb_id.clone(),
        media_type: body.media_type.clone(),
        season: body.season,
        episode: body.episode,
        title: body.title,
        poster_url: body.poster_url,
        magnet_uri: Some(body.magnet_uri),
        infohash: Some(body.infohash),
        torrent_name: Some(body.torrent_name),
        file_idx: Some(body.file_idx),
        file_size_bytes: Some(body.file_size_bytes),
    };

    queries::insert_job(&state.db, &new_job).await?;

    // Log event
    queries::insert_job_event(
        &state.db, &job_id, None, "status_change",
        "Job queued", None,
    ).await?;

    // Broadcast
    let _ = state.event_tx.send(crate::api::events::SseEvent::JobCreated {
        job_id: job_id.clone(),
        title: new_job.title.clone().unwrap_or_default(),
    });

    // Telegram notification
    notifications::send_notification(&state, TelegramEvent::JobQueued(
        new_job.title.clone().unwrap_or_default(),
    ));

    Ok(Json(CreateJobResponse {
        job_id,
        status: "queued".to_string(),
    }))
}

pub async fn list_jobs(
    State(state): State<Arc<AppState>>,
) -> AppResult<Json<QueueListResponse>> {
    let all_jobs = queries::list_jobs(&state.db).await?;

    let mut processing = Vec::new();
    let mut queued = Vec::new();
    let mut completed = Vec::new();
    let mut failed = Vec::new();

    let processing_statuses = [
        "processing", "downloading", "checkpoint_download",
        "transcoding", "checkpoint_transcode", "uploading",
    ];

    for job in all_jobs {
        match job.status.as_str() {
            "queued" => queued.push(job),
            "completed" => completed.push(job),
            "failed" => failed.push(job),
            s if processing_statuses.contains(&s) => processing.push(job),
            _ => queued.push(job), // fallback
        }
    }

    Ok(Json(QueueListResponse { processing, queued, completed, failed }))
}
pub async fn get_job(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
) -> AppResult<Json<JobDetailResponse>> {
    let job = queries::get_job(&state.db, &id).await
        .map_err(|_| AppError::NotFound(format!("Job {} not found", id)))?;
    let events = queries::get_job_events(&state.db, &id).await?;
    let gh_repo = get_setting_or_env(&state, "gh_repo").await?.filter(|repo| !repo.is_empty());

    Ok(Json(JobDetailResponse { job, events, gh_repo }))
}

pub async fn retry_job(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
) -> AppResult<Json<RetryResponse>> {
    let job = queries::get_job(&state.db, &id).await
        .map_err(|_| AppError::NotFound(format!("Job {} not found", id)))?;

    if job.status != "failed" {
        return Err(AppError::BadRequest("Can only retry failed jobs".into()));
    }

    let skip_download = job.last_checkpoint.as_deref() == Some("download")
        || job.last_checkpoint.as_deref() == Some("transcode");
    let skip_transcode = job.last_checkpoint.as_deref() == Some("transcode");

    queries::insert_job_event(
        &state.db, &id, None, "status_change",
        &format!("Retry triggered (last checkpoint: {:?}, skip_dl: {}, skip_tc: {})",
            job.last_checkpoint, skip_download, skip_transcode),
        None,
    ).await?;

    // Trigger pipeline with skip flags instead of resetting to queued
    match trigger::trigger_pipeline(&state, &job, skip_download, skip_transcode).await {
        Ok(run_id) => {
            queries::update_job_status(&state.db, &id, "processing").await?;
            queries::insert_job_event(
                &state.db, &id, None, "status_change",
                &format!("Retry pipeline triggered (run_id: {})", run_id), None,
            ).await?;

            let _ = state.event_tx.send(crate::api::events::SseEvent::JobRetried {
                job_id: id.clone(),
            });

            Ok(Json(RetryResponse {
                job_id: id,
                status: "processing".to_string(),
            }))
        }
        Err(e) => Err(e),
    }
}

pub async fn delete_job(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
) -> AppResult<Json<serde_json::Value>> {
    let job = queries::get_job(&state.db, &id).await
        .map_err(|_| AppError::NotFound(format!("Job {} not found", id)))?;

    let active_statuses = [
        "processing", "downloading", "transcoding", "uploading",
        "checkpoint_download", "checkpoint_transcode",
    ];
    let is_active = active_statuses.contains(&job.status.as_str());

    // Cancel GitHub Actions run if the job is active
    if is_active {
        if let Some(ref run_id) = job.gh_run_id {
            if run_id != "pending" && !run_id.is_empty() {
                let config = state.config.read().await;
                let gh_token = crate::pipeline::trigger::get_setting_or_env(&state, "gh_token").await?.unwrap_or_default();
                let gh_repo = crate::pipeline::trigger::get_setting_or_env(&state, "gh_repo").await?.unwrap_or_default();
                drop(config);

                if !gh_token.is_empty() && !gh_repo.is_empty() {
                    match crate::pipeline::trigger::cancel_gh_run(&state.http, &gh_token, &gh_repo, run_id).await {
                        Ok(()) => {
                            eprintln!("[queue] Cancelled GH run {run_id} for job {id}");
                            queries::insert_job_event(
                                &state.db, &id, None, "status_change",
                                &format!("Cancelled GitHub Actions run {run_id}"), None,
                            ).await?;
                        }
                        Err(e) => {
                            eprintln!("[queue] Failed to cancel GH run {run_id}: {e}");
                            // Non-fatal — still proceed with DB deletion
                        }
                    }
                }
            }
        }
    }

    // Delete from DB (cascades to job_events and hls_chunks)
    queries::delete_job(&state.db, &id).await?;

    // Broadcast SSE event
    let _ = state.event_tx.send(crate::api::events::SseEvent::JobRemoved {
        job_id: id.clone(),
    });

    Ok(Json(serde_json::json!({ "removed": true, "cancelled_run": is_active })))
}

File: backend-bun/src/api/queue.ts
import { toSnake } from "../db/transform";
import type { Context } from "hono";
import type { AppBindings } from "../app";
import { badRequest } from "../error";
import * as queries from "../db/queries";
import { triggerPipeline, cancelGhRun } from "../pipeline/trigger";
import { sendNotification } from "../notifications/telegram";
import { buildMagnet } from "./search";
import { eq, sql } from "drizzle-orm";
import { jobs } from "../db/schema";

const ACTIVE_STATUSES = [
  "processing",
  "downloading",
  "transcoding",
  "uploading",
];

export async function createJob(c: Context<AppBindings>) {
  const raw = await c.req.json();
  const body = raw as Record<string, unknown>;
  // Required fields — match Rust CreateJobRequest validation
  const imdbId = body.imdb_id;
  const mediaType = body.media_type;
  const infohash = body.infohash;
  const torrentName = (body.torrent_name as string) ?? "";
  const title = (body.title as string) ?? torrentName;

  if (!imdbId || typeof imdbId !== "string" || !imdbId.trim()) throw badRequest("imdb_id is required");
  if (!mediaType || typeof mediaType !== "string" || !mediaType.trim()) throw badRequest("media_type is required");
  if (!infohash || typeof infohash !== "string" || !infohash.trim()) throw badRequest("infohash is required");

  // Auto-construct magnet URI with trackers if not provided (backward compat)
  const magnetUri = (body.magnet_uri as string)?.trim()
    || buildMagnet(infohash, title);
  // ponytail: validate custom magnet URIs to prevent injection into CI pipelines
  if (magnetUri && !magnetUri.startsWith("magnet:?")) {
    throw badRequest("Invalid magnet URI format — must start with magnet:?");
  }

  const jobId = crypto.randomUUID();

  const newJob: queries.NewJob = {
    id: jobId,
    imdbId: imdbId,
    mediaType: mediaType,
    season: (body.season as number | undefined) ?? null,
    episode: (body.episode as number | undefined) ?? null,
    title: title || null,
    posterUrl: (body.poster_url as string | undefined) ?? null,
    magnetUri: magnetUri,
    infohash: infohash,
    torrentName: torrentName || null,
    fileIdx: (body.file_idx as number) ?? 0,
    fileSizeBytes: (body.file_size_bytes as number) ?? 0,
  };

  queries.insertJob(c.var.db, newJob);
  queries.insertJobEvent(c.var.db, jobId, null, "status_change", "Job queued", null);

  const eventTitle = newJob.title ?? "";
  c.var.eventBus.send({ type: "job_created", data: { job_id: jobId, title: eventTitle } });
  sendNotification(c, { type: "JobQueued", title: eventTitle });

  return c.json({ job_id: jobId, status: "queued" }, 201);
}

export async function listJobs(c: Context<AppBindings>) {
  const allJobs = queries.listJobs(c.var.db);

  const processing: queries.Job[] = [];
  const queued: queries.Job[] = [];
  const completed: queries.Job[] = [];
  const failed: queries.Job[] = [];

  for (const job of allJobs) {
    if (ACTIVE_STATUSES.includes(job.status)) {
      processing.push(job);
    } else if (job.status === "queued") {
      queued.push(job);
    } else if (job.status === "completed") {
      completed.push(job);
    } else if (job.status === "failed") {
      failed.push(job);
    } else {
      // ponytail: unknown status fallback — treat as queued
      queued.push(job);
    }
  }
  return c.json({ processing: toSnake(processing), queued: toSnake(queued), completed: toSnake(completed), failed: toSnake(failed) });
}

export async function getJob(c: Context<AppBindings>) {
  const id = c.req.param("id") as string;
  const job = queries.getJob(c.var.db, id);
  const events = queries.getJobEvents(c.var.db, id);
  const ghRepo = queries.getSetting(c.var.db, "gh_repo") || c.var.config.ghRepo || null;

  return c.json({ job: toSnake(job), events: toSnake(events), gh_repo: ghRepo });
}

export async function retryJob(c: Context<AppBindings>) {
  const id = c.req.param("id") as string;
  const job = queries.getJob(c.var.db, id);

  if (job.status !== "failed") {
    throw badRequest("Can only retry failed jobs");
  }
  // Auto-construct magnetUri for old jobs created before validation fix
  if (!job.magnetUri?.trim() && job.infohash) {
    const dn = job.title || job.torrentName || job.imdbId;
    job.magnetUri = buildMagnet(job.infohash, dn ?? "unknown");
    // Update DB so future retries and page loads show the magnet URI
    c.var.db.update(jobs)
      .set({ magnetUri: job.magnetUri, updatedAt: sql`(datetime('now'))` })
      .where(eq(jobs.id, id))
      .run();
  }
  try {
    const runId = await triggerPipeline(c, job);
    queries.updateJobStatus(c.var.db, id, "processing");
    queries.insertJobEvent(
      c.var.db,
      id,
      null,
      "status_change",
      `Retry pipeline triggered (run_id: ${runId})`,
      null,
    );
    c.var.eventBus.send({ type: "job_retried", data: { job_id: id } });
    return c.json({ job_id: id, status: "processing" });
  } catch (e) {
    throw badRequest((e as Error).message);
  }
}

export async function deleteJobHandler(c: Context<AppBindings>) {
  const id = c.req.param("id") as string;
  const job = queries.getJob(c.var.db, id);

  const isActive = ACTIVE_STATUSES.includes(job.status);

  if (isActive && job.ghRunId && job.ghRunId !== "pending") {
    const ghToken = queries.getSetting(c.var.db, "gh_token") || c.var.config.ghToken || "";
    const ghRepo = queries.getSetting(c.var.db, "gh_repo") || c.var.config.ghRepo || "";
    if (ghToken && ghRepo) {
      try {
        await cancelGhRun(ghToken, ghRepo, job.ghRunId);
      } catch (e) {
        console.error(`[queue] Failed to cancel GH run ${job.ghRunId}:`, (e as Error).message);
      }
    }
  }

  queries.deleteJob(c.var.db, id);
  c.var.eventBus.send({ type: "job_removed", data: { job_id: id } });

  return c.json({ removed: true, cancelled_run: isActive });
}
